src/goSuggest.ts

```diff
@@ -533,3 +473,137 @@ export class GoCompletionItemProvider implements vscode.CompletionItemProvider,
+		currentWord = word.substr(0, position.character - wordAtPosition.start.character);
+	}
+
+	return currentWord;
+}
+
+function getKeywordCompletions(currentWord: string): vscode.CompletionItem[] {
+	if (!currentWord.length) {
+		return [];
+	}
+	const completionItems: vscode.CompletionItem[] = [];
+	goKeywords.forEach(keyword => {
+		if (keyword.startsWith(currentWord)) {
+			completionItems.push(new vscode.CompletionItem(keyword, vscode.CompletionItemKind.Keyword));
+		}
+	});
+	return completionItems;
+}
+
+/**
+ * Return importable packages that match given word as Completion Items
+ * @param document Current document
+ * @param currentWord The word at the cursor
+ * @param allPkgMap Map of all available packages and their import paths
+ * @param importedPackages List of imported packages. Used to prune imported packages out of available packages
+ */
+function getPackageCompletions(document: vscode.TextDocument, currentWord: string, allPkgMap: Map<string, string>, importedPackages: string[] = []): vscode.CompletionItem[] {
+	const cwd = path.dirname(document.fileName);
+	const goWorkSpace = getCurrentGoWorkspaceFromGOPATH(getCurrentGoPath(), cwd);
+	const workSpaceFolder = vscode.workspace.getWorkspaceFolder(document.uri);
+	const currentPkgRootPath = (workSpaceFolder ? workSpaceFolder.uri.path : cwd).slice(goWorkSpace.length + 1);
+
+	const completionItems: any[] = [];
+
+	allPkgMap.forEach((pkgName: string, pkgPath: string) => {
+		if (pkgName.startsWith(currentWord) && importedPackages.indexOf(pkgName) === -1) {
+
+			const item = new vscode.CompletionItem(pkgName, vscode.CompletionItemKind.Keyword);
+			item.detail = pkgPath;
+			item.documentation = 'Imports the package';
+			item.insertText = pkgName;
+			item.command = {
+				title: 'Import Package',
+				command: 'go.import.add',
+				arguments: [{ importPath: pkgPath, from: 'completion' }]
+			};
+			item.kind = vscode.CompletionItemKind.Module;
+
+			// Unimported packages should appear after the suggestions from gocode
+			const isStandardPackage = !item.detail.includes('.');
+			item.sortText = isStandardPackage ? 'za' : pkgPath.startsWith(currentPkgRootPath) ? 'zb' : 'zc';
+			completionItems.push(item);
+		}
+	});
+	return completionItems;
+}
+
+export async function getCompletionsWithoutGoCode(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.CompletionItem[]> {
+	const lineText = document.lineAt(position.line).text;
+	const config = vscode.workspace.getConfiguration('go', document.uri);
+	const autocompleteUnimportedPackages = config['autocompleteUnimportedPackages'] === true && !lineText.match(/^(\s)*(import|package)(\s)+/);
+
+	const commentCompletion = getCommentCompletion(document, position);
+	if (commentCompletion) {
+		return [commentCompletion];
+	}
+
+	if (isPositionInComment(document, position)) {
+		return [];
+	}
+
+	const currentWord = getCurrentWord(document, position);
+	if (!currentWord.length) {
```

Comment:
>!currentWord.length [](start = 5, length = 19)

is this supposed to check that length exists ? or if its (0) ? if its the later, maybe if we can compare explicitly to avoid confusions ?

Reply:
seems that getCurrentWorkd always returns a string so I assume is the later ?

---
In reply to: [270206674](https://github.com/Microsoft/vscode-go/pull/2383#discussion_r270206674) [](ancestors = 270206674)

Reply:
In this case, `!currentWord.length` is a like a shorter check for `currentWord.length === 0`
I am ensuring that the word is a non empty string.